### DISCUSSION TITLE
DiscreteInput requires bounds Error

### DISCUSSION THREAD

**Discussion by behradesg** [experimental-design/bofire#422] [Q&A]
Dear All,

I have been trying out one of the examples of DOE using DiscreteInput and I am encountering the following error:
"ValueError: `bounds` must specify both lower and upper limits for each decision variable."
Here is my code:
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.ticker import FormatStrFormatter

from bofire.data_models.constraints.api import (
    NonlinearEqualityConstraint,
    NonlinearInequalityConstraint,
    LinearEqualityConstraint,
    LinearInequalityConstraint,
    InterpointEqualityConstraint,
)
from bofire.data_models.domain.api import Domain
from bofire.data_models.features.api import ContinuousInput, ContinuousOutput, DiscreteInput
from bofire.strategies.doe.design import find_local_max_ipopt

domain = Domain(
   inputs = [
    DiscreteInput(key="x1", values=[i for i in range(400, 850, 50)]), 
    DiscreteInput(key="x2", values=[0, 0.025, 0.05, 0.075, 0.1]),
    DiscreteInput(key="x3", values=[1, 2, 3])
    ],
   outputs = [ContinuousOutput(key="y")]
)
I tried to specify the bounds as well, but I encounter another error:
"ValidationError: 1 validation error for DiscreteInput
bounds
  Extra inputs are not permitted [type=extra_forbidden, input_value=(400, 850), input_type=tuple]
    For further information visit https://errors.pydantic.dev/2.8/v/extra_forbidden"
   
   I would highly appreciate your help in this regard. Looking forward to hearing from you.
   
   Thanks,
   Behrad

**Comment by Osburg** [comment]
Hi Behrad :),

Unfortunately I cannot reproduce your first error from the code you posted. From your code I assumed that you are referring to the first example in [this notebook](https://github.com/experimental-design/bofire/blob/main/tutorials/doe/basic_examples.ipynb). 

When I run the code from this example replacing the original domain definition with your discrete input example the error i get is a different one
`ValueError: lb and ub must either be None or have length n.`

This happens since `find_local_max_ipopt()` only supports continuous inputs.
You can check out `find_local_max_ipopt_BaB()` and `find_local_max_ipopt_exhaustive()` if you want to use categorical inputs.

Cheers
Aaron

**Reply by behradesg** [reply]
Hi Aaron,

Thanks for your reply. It is strange that you cannot produce the same error. Anyway, I switched to find_local_max_ipopt_exhaustive() but I get another error now. Here is the code: 
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.ticker import FormatStrFormatter

from bofire.data_models.constraints.api import (
    NonlinearEqualityConstraint,
    NonlinearInequalityConstraint,
    LinearEqualityConstraint,
    LinearInequalityConstraint,
    InterpointEqualityConstraint,
)
from bofire.data_models.domain.api import Domain
from bofire.data_models.features.api import ContinuousInput, ContinuousOutput, DiscreteInput
from bofire.strategies.doe.design import find_local_max_ipopt_exhaustive

domain = Domain(
   inputs = [
    DiscreteInput(key="x1", values=[400, 450, 500, 550, 600, 650, 700, 750, 800] ), 
    DiscreteInput(key="x2", values=[0, 0.025, 0.05, 0.075, 0.1]),
    DiscreteInput(key="x3", values=[1, 2, 3])
    ],
   outputs = [ContinuousOutput(key="y")]
)
d_optimal_design = find_local_max_ipopt_exhaustive(domain, "linear").to_numpy().T

Here is the error I get: 
"
  File ~/anaconda3/envs/cyipopt-env/lib/python3.12/site-packages/bofire/strategies/doe/design.py:221 in find_local_max_ipopt_exhaustive
    if discrete_variables is not None or len(discrete_variables) > 0:  # type: ignore

TypeError: object of type 'NoneType' has no len() "

Basically what I want to do is very simple, I have three discrete parameters and I want to get the experimental combinations of them. Would be possible for you to help me to do this simple task? 

Thank you so much!
Behrad

**Reply by Osburg** [reply]
Sure, I can try to help you. Are you interested in a list of all possible experimental combinations or in a (d-)optimal experimental design with a given number of experiments `n_experiment`? In the latter case: what is `n_experiments` in your situation?

**Reply by behradesg** [reply]
Dear Aaron,

Thank you so much. I am only interested in a list of all possible experimental combinations.

**Reply by Osburg** [reply]
Dear Behrad, 

I think you get what you want to have with the following lines of code:

```
import itertools

x1 = [400, 450, 500, 550, 600, 650, 700, 750, 800]
x2 = [0, 0.025, 0.05, 0.075, 0.1]
x3 = [1, 2, 3]

all_combinations = list(itertools.product(x1, x2, x3))
```

Does this solve your problem?

**Reply by behradesg** [reply]
Thank you so much. This is what I was looking for.

**Reply by Osburg** [reply]
Sure, no problem :)
